pkg/storage/stores/shipper/compactor/deletion/delete_request.go

```diff
@@ -15,21 +17,60 @@ type DeleteRequest struct {
 	Status    DeleteRequestStatus `json:"status"`
 	CreatedAt model.Time          `json:"created_at"`
 
-	UserID   string            `json:"-"`
-	matchers []*labels.Matcher `json:"-"`
+	UserID          string                 `json:"-"`
+	matchers        []*labels.Matcher      `json:"-"`
+	logSelectorExpr syntax.LogSelectorExpr `json:"-"`
 }
 
 func (d *DeleteRequest) SetQuery(logQL string) error {
 	d.Query = logQL
-	matchers, err := parseDeletionQuery(logQL)
+	logSelectorExpr, err := parseDeletionQuery(logQL)
 	if err != nil {
 		return err
 	}
-	d.matchers = matchers
+	d.logSelectorExpr = logSelectorExpr
+	d.matchers = logSelectorExpr.Matchers()
 	return nil
 }
 
-func (d *DeleteRequest) IsDeleted(entry retention.ChunkEntry) (bool, []model.Interval) {
+// FilterFunction returns a filter function that returns true if the given line matches
+func (d *DeleteRequest) FilterFunction(labels labels.Labels) (filter.Func, error) {
+	if d.logSelectorExpr == nil {
+		err := d.SetQuery(d.Query)
+		if err != nil {
+			return nil, err
+		}
+	}
+	p, err := d.logSelectorExpr.Pipeline()
+	if err != nil {
+		return nil, err
+	}
+
+	f := p.ForStream(labels).ProcessString
+	return func(s string) bool {
+		if !allMatch(d.matchers, labels) {
+			return false
+		}
+		result, _, skip := f(s)
+		if len(result) != 0 || skip {
```

Comment:
we can just `return len(result) != 0 || skip`

Reply:
Indeed.